rfcs/0180-broken-package-removal.md

```diff
@@ -0,0 +1,99 @@
+---
+feature: broken-package-removal
+start-date: 2024-07-14
+author: Jörg "Mic92" Thalheim
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+---
+
+# Summary
+
+[summary]: #summary
+
+This RFCs defines under what conditions we remove broken and unmaintained leaf packages.
+The RFC does not have the ambition to define all reasons to remove a package but
+instead focuses on simple rules that can be automated.
+Because this document is just the guideline for automatic package removal,
+it also leave it open, that a package can be removed earlier for other reasons.
+
+# Motivation
+
+[motivation]: #motivation
+
+Broken and unmaintained packages still cost us valuable time.
+Broken packages still have to be evaluated to some extent
+and unmaintained packages might still have to rebuild when one of their dependencies changes.
+Removing unmaintained/unused code hopefully increases the overall package quality and helps us
+to save resources.
+Because it is easy to add packages, it should be also easy to remove them, ideally automated.
+
+# Detailed design
+
+[design]: #detailed-design
+
+All packages marked as broken/unmaintained will be removed after a full NixOS release.
+For example `hello-unmaintained` will be marked as unmaintained in 23.11 then once we
+have 24.05, we can remove it from master.
+
+## Broken Packages
+
+Packages that are marked as broken unconditionally platforms should are canidates for removal.
+If the package had dependencies, those dependencies are also marked as indirectly broken.
+If they are not functional without the broken package, they should be removed as well.
+
+## Unmaintained packages
+
+We add an unmaintained warning for all packages, with an empty maintainer field that do not have packages depending on it.
+Ideally we have an automation or semi-automation that creates pull requests for that (see future work).
+
+If a NixOS module depends on any of removed package, this module gets removed as well, if it is non-functional without the package.
+In `pkgs/top-level/aliases.nix` we can link to the pull request that removed the package,
+to make it easier for people to recover the nix expression.
+
+In the pull request that removes a package, we also will ping people that have modified/updated the package excluding those that only
+touched the package as part of treewide.
+
+# Examples and Interactions
+
+[examples-and-interactions]: #examples-and-interactions
+
+-
```

Comment:
```suggestion
- Release manager or other contributior marks `packageA` as broken.
- Automated PR will be opened by bot.
- Notify Maintainer(s) they have one full release cycle as described to fix the package.
- If the package is not fixed it will be removed as described, by merging the PR.
- If the package is fixed. the PR is closed automatically without merging.
```

Reply:
Slightly adapted and added.